game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/battlegroup/NormalBattleGroups.java

```diff
@@ -0,0 +1,159 @@
+                            .test(targetUnitFilterData.getTargetUnit())))
+        .enemyUnitRequirements(enemyUnitPredicate);
+  }
+
+  private Predicate<FiringSquadron.FiringUnitFilterData> unitIsFirstStrike(
+      final GameProperties properties) {
+    return firingUnitFilterData ->
+        PredicateBuilder.<Unit>trueBuilder()
+            .andIf(firingUnitFilterData.getSide() == OFFENSE, Matches.unitIsFirstStrike())
+            .andIf(
+                firingUnitFilterData.getSide() == DEFENSE,
+                Matches.unitIsFirstStrikeOnDefense(properties))
+            .build()
+            .test(firingUnitFilterData.getFiringUnit());
+  }
+
+  private Predicate<Collection<Unit>> noDestroyerPresent() {
+    return units -> units.stream().noneMatch(Matches.unitIsDestroyer());
+  }
+
+  /**
+   * Creates battle groups for the general phase
+   *
+   * <p>Since some units may fire during the first strike phase depending on the presence of an
+   * isDestroyer, this creates two sets of FiringSquadrons. One set includes all units but only
+   * fires if there is an isDestroyer, the other set only includes non isFirstStrike units but only
+   * fires if there is not an isDestroyer.
+   */
+  static BattleGroup createGeneral(
+      final Collection<UnitType> unitTypes, final GameProperties properties) {
+
+    final List<FiringSquadron> firingSquadronsWithoutFirstStrikeUnits =
+        FiringSquadron.createWithTargetInformation(unitTypes).stream()
+            .map(
+                firingSquadron -> {
+                  return createFiringSquadron(
+                          firingSquadron,
+                          Predicate.not(unitIsFirstStrike(properties)),
+                          noDestroyerPresent())
+                      .name(UNITS)
+                      .build();
+                })
+            .collect(Collectors.toList());
+
+    final List<FiringSquadron> firingSquadronsWithFirstStrikeUnits =
+        FiringSquadron.createWithTargetInformation(unitTypes).stream()
+            .map(
+                firingSquadron -> {
+                  return createFiringSquadron(
+                          firingSquadron,
+                          firingUnitFilterData -> true,
+                          Predicate.not(noDestroyerPresent()))
+                      .name(UNITS)
+                      .build();
+                })
+            .collect(Collectors.toList());
+
+    final List<FiringSquadron> offenseSquadrons =
+        Stream.of(firingSquadronsWithFirstStrikeUnits, firingSquadronsWithoutFirstStrikeUnits)
+            .flatMap(Collection::stream)
+            .collect(Collectors.toList());
+
+    final List<FiringSquadron> defenseSquadrons;
+    if (Properties.getDefendingSubsSneakAttack(properties)) {
+      defenseSquadrons = offenseSquadrons;
+    } else {
+      defenseSquadrons =
+          FiringSquadron.createWithTargetInformation(unitTypes).stream()
+              .map(
+                  firingSquadron -> {
```

Comment:
nit, this could be written with the simplified lambda syntax?
I wonder as well if more of this logic couldn't be rolled into `FiringSquadron`. There almost seems to be a feature envy code-smell here. Notably as well I would take a look at the `createFiringSquadron` that accepts predicates. I am starting to think that any predicate parameter is an outright code-smell (like a boolean parameter).

Reply:
See https://github.com/triplea-game/triplea/pull/8602/commits/bb56e98c44c6efaec443b627729f2627d646961f